imagebuildah/build.go

```diff
@@ -340,6 +340,32 @@ func buildDockerfilesOnce(ctx context.Context, store storage.Store, logger *logr
 
 	warnOnUnsetBuildArgs(logger, mainNode, options.Args)
 
+	// --platform was explicitly selected for this build
+	// so set correct TARGETPLATFORM in args if it is not
+	// already selected by the user.
+	if options.SystemContext.OSChoice != "" && options.SystemContext.ArchitectureChoice != "" {
+		// os component from --platform string populates TARGETOS
+		if _, ok := options.Args["TARGETOS"]; !ok {
+			options.Args["TARGETOS"] = options.SystemContext.OSChoice
+		}
+		// arch component from --platform string populates TARGETARCH
+		if _, ok := options.Args["TARGETARCH"]; !ok {
+			options.Args["TARGETARCH"] = options.SystemContext.ArchitectureChoice
+		}
+		// variant component from --platform string populates TARGETVARIANT
+		if _, ok := options.Args["TARGETVARIANT"]; !ok {
+			if options.SystemContext.VariantChoice != "" {
+				options.Args["TARGETVARIANT"] = options.SystemContext.VariantChoice
+			}
+		}
+		if _, ok := options.Args["TARGETPLATFORM"]; !ok {
+			options.Args["TARGETPLATFORM"] = options.SystemContext.OSChoice + "/" + options.SystemContext.ArchitectureChoice
```

Comment:
Shouldn't this be:
options.Args["TARGETPLATFORM"] = options.Args["TARGETOS"] + "/" + options.Args["TARGETARCH"]

Reply:
If users wants he can override `TARGETOS` and `TARGETARGS` so there can be scenario when `options.Args["TARGETOS"] ` and `options.Args["TARGETARCH"]` are not set.

Reply:
IMO I think in the end they just `build-args` so user should always have ability to override manually any one of them without disturbing the defaults of other `build-args`. So this ensures that `TARGETPLATFORM` will always point to correct default if it was not meant to be overridden.

Reply:
If user set options.SystemContext.OSChoice then it will be the same in options.Args["TARGETPLATFORM"]  via the code above.  If they did not set it, then you will end with TARGETPLATFORM =="", potentially.

Reply:
I think the line `if _, ok := options.Args["TARGETOS"]; !ok {` verifies that user did not use `--build-arg TARGETOS=something` so value of `options.Args["TARGETOS"]` will be `options.SystemContext.OSChoice` only if `--build-arg` was not used

```go
	if _, ok := options.Args["TARGETOS"]; !ok {
                        // only set when user did not use --build-arg TARGETOS=something
			options.Args["TARGETOS"] = options.SystemContext.OSChoice
	}
```
so if we do 
`options.Args["TARGETPLATFORM"] = options.Args["TARGETOS"] + "/" + options.Args["TARGETARCH"]` and user does
`--build-arg TARGETOS=android` then `TARGETPLATFORM` becomes `android/<somearch>` which seems wrong.

I have verified this behaviour with buildkit as well here is the output with test `Containerfile` included in this PR
```console
sudo docker buildx build --platform=linux/arm64 --no-cache --progress=plain --build-arg TARGETOS=android -t test .
```
```console
...

#5 [2/2] RUN echo "I'm compiling for linux/arm64 on linux/amd64 and tagging for linux/arm64 and OS android and ARCH arm64"
#5 0.311 I'm compiling for linux/arm64 on linux/amd64 and tagging for linux/arm64 and OS android and ARCH arm64
#5 DONE 0.4s
...
```

also added a test in CI which ensures we don't regress this.

Reply:
So accepting this change would break test `build-with-inline-platform-and-rely-on-defaultbuiltinargs-check-custom-override` in this PR.

Reply:
We discussed this and keeping behviour as is since it matches with buildkit but can change if this is found to be bug in buildkit.